### PR TITLE
ES-421

### PR DESCRIPTION
**Pull request by ase-101**
1. AddIdentity  endpoint - based on the selectedHandles, creates handle-uin mapping and handle credential issued to IDA.
2. Added endpoint to retrieve handles using uin.
3. Added new IdType - handle
4. Test case fixes
5. DB script changes for new "handle" table.

### CODE REVIEW

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+            if(identityMap.containsKey(SELECTED_HANDLES) && Objects.nonNull(identityMap.get(SELECTED_HANDLES))) {
+                mosipLogger.debug(IdRepoSecurityManager.getUser(), ID_REPO_SERVICE_HELPER, "getSelectedHandles",
+                        requestMap.get(SELECTED_HANDLES));
+                List<String> selectedHandleFieldIds = (List<String>) identityMap.get(SELECTED_HANDLES);
+                return selectedHandleFieldIds.stream()
+                        .filter( handleFieldId -> supportedHandlesInSchema.get(schemaVersion).contains(handleFieldId))
+                        .collect(Collectors.toMap(handleName->handleName,
+                                handleFieldId-> {
+                                    String handle = ((String) identityMap.get(handleFieldId))
+                                            .concat("@").concat(handleFieldId)
```
Comment (vishwa-vyom) [inline id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java:135]: We should have a map between fieldId and the actual handle type that can be used. Please create a JIRA task

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            throw new IdRepoAppUncheckedException(IdRepoErrorConstants.SCHEMA_RETRIEVE_ERROR);
+        }
+    }
+
+    public Map<String, HandleDto> getSelectedHandles(Object identity) throws IdRepoAppException {
+        Map<String, Object> requestMap = convertToMap(identity);
+        if (requestMap.containsKey(ROOT_PATH) && Objects.nonNull(requestMap.get(ROOT_PATH))) {
+            Map<String, Object> identityMap = (Map<String, Object>) requestMap.get(ROOT_PATH);
+            String schemaVersion = String.valueOf(identityMap.get(ID_SCHEMA_VERSION));
+            if(identityMap.containsKey(SELECTED_HANDLES) && Objects.nonNull(identityMap.get(SELECTED_HANDLES))) {
```
Comment (vishwa-vyom) [inline id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java:126]: keys for id schema and selected handles need to be taken from identity mapping and then queried into the identity object

File: db_release_scripts/mosip_idrepo/sql/1.2.0_release.sql [hunk 3]
```diff
@@ -29,6 +29,7 @@ ALTER TABLE idrepo.uin_auth_lock ADD COLUMN unlock_expiry_datetime timestamp;
 \ir ../ddl/idrepo-anonymous_profile.sql
 \ir ../ddl/idrepo-channel_info.sql
 \ir ../ddl/idrepo-identity_update_count_tracker.sql
+\ir ../ddl/idrepo-handle.sql
```
[the remark this answers is not in the document]
Comment (ase-101) [reply db_release_scripts/mosip_idrepo/sql/1.2.0_release.sql:32]: Renamed to 1.2.1

File: db_release_scripts/mosip_idrepo/sql/1.2.0_revoke.sql [hunk 4]
```diff
@@ -28,4 +28,8 @@ DROP TABLE IF EXISTS idrepo.anonymous_profile;
 DROP TABLE IF EXISTS idrepo.channel_info;
 ------------------------------------------------------------------------------------------------------
 
-DROP TABLE IF EXISTS idrepo.idrepo-identity_update_count_tracker;
\ No newline at end of file
+DROP TABLE IF EXISTS idrepo.identity_update_count_tracker;
+
+------------------------------------------------------------------------------------------------------
+
+DROP TABLE IF EXISTS idrepo.handle;
```
[the remark this answers is not in the document]
Comment (ase-101) [reply db_release_scripts/mosip_idrepo/sql/1.2.0_revoke.sql:35]: Renamed to 1.2.1

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/dto/HandleInfoDTO.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class HandleInfoDTO {
+
+    private String handle;
+    private LocalDateTime expiryTimestamp;
+    private LocalDateTime genratedOnTimestamp;
+    private Integer transactionLimit;
+    private Map<String, String> hashAttributes;
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/dto/HandleInfoDTO.java:19]: Done

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/dto/HandleInfosDTO.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+import io.mosip.kernel.core.http.ResponseWrapper;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.util.List;
+
+@Data
+@EqualsAndHashCode(callSuper=true)
+public class HandleInfosDTO extends ResponseWrapper<List<HandleInfoDTO>> {
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/dto/HandleInfosDTO.java:11]: Removed

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java [hunk 7]
```diff
@@ -190,9 +180,16 @@ public void notifyUinCredential(String uin, LocalDateTime expiryTimestamp, Strin
 				sendUINEventToIDA(uin, expiryTimestamp, status, vidInfoDtos, partnerIds, txnId,
 						id -> securityManager.getIdHashWithSaltModuloByPlainIdHash(id, saltRetreivalFunction), idaEventModelConsumer);
 			} else {
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java:205]: Added mosip.idrepo.vid.enabled flag.

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java [hunk 8]
```diff
@@ -362,6 +360,17 @@ public void sendUinEventsToCredService(String uin, LocalDateTime expiryTimestamp
 			eventRequestsList.addAll(vidRequests);
 		}
 
+		if(handleList != null) {
+			List<CredentialIssueRequestDto> handleRequests = handleList.stream().flatMap(handleInfoDTO -> {
+				return partnerIds.stream().map(partnerId -> {
+					String token = tokenIDGenerator.generateTokenID(uin, partnerId);
+					return createCredReqDto(handleInfoDTO.getHandle(), partnerId, null, null,
+							token, handleInfoDTO.getHashAttributes(), requestId);
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java]: Done

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/entity/Handle.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import java.time.LocalDateTime;
+
+@Data
+@Entity
+@NoArgsConstructor
+@Table(name = "handle", schema = "idrepo")
+public class Handle implements HandleInfo {
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/entity/Handle.java]: Done

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/repository/HandleRepo.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package io.mosip.idrepository.identity.repository;
+
+import io.mosip.idrepository.identity.entity.Handle;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.List;
+
+public interface HandleRepo extends JpaRepository<Handle, String> {
```
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/repository/HandleRepo.java]: Done

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java (hunk 1 above)
Comment (ase-101) [reply id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java:158]: https://mosip.atlassian.net/browse/ES-572

File: id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java (hunk 2 above)
Comment (ase-101) [reply id-repository/id-repository-identity-service/src/main/java/io/mosip/idrepository/identity/helper/IdRepoServiceHelper.java]: Done

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java:205]: Normally the default values for boolean to be false, can we change this to mosip.idrepo.vid.disable-support

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vishwa-vyom) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java]: We have to document this very clearly somewhere, do we have a confluence page with the approch details for us to later move to md files ? Also can we detail out under the convention based request id config in the property files ?

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java [hunk 11]
```diff
@@ -565,4 +592,31 @@ private CredentialStatusUpdateEvent createCredentialStatusUpdateEvent(String req
...
+							.retrieveSaltById(Integer.valueOf(io.mosip.kernel.core.util.StringUtils.substringBefore(entity.getHandle(), SPLITTER)));
+					try {
+						handleInfoDTO.setHandle(new String(securityManager.decryptWithSalt(
+								CryptoUtil.decodeURLSafeBase64(io.mosip.kernel.core.util.StringUtils.substringAfter(entity.getHandle(), SPLITTER)),
+								CryptoUtil.decodePlainBase64(encryptSalt), uinRefId)));
+					} catch (IdRepoAppException e) {
+						mosipLogger.error(IdRepoSecurityManager.getUser(), SEND_REQUEST_TO_CRED_SERVICE, "getHandles",
+								"\n Failed to decrypt handle due to " + e.getMessage());
+					}
+					handleInfoDTO.setAdditionalData(securityManager.getIdHashAndAttributesWithSaltModuloByPlainIdHash(entity.getHandle(),
```
Comment (vishwa-vyom) [inline id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java:615]: what are we setting here ?

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java]: ok

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java (hunk 11 above)
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java:615]: saltId, module and idHash, its an existing method.

File: id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ase-101) [reply id-repository/id-repository-core/src/main/java/io/mosip/idrepository/core/manager/CredentialServiceManager.java:205]: changed the property name.

File: db_release_scripts/mosip_credential/sql/1.2.1_release.sql [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+-- Database Name: mosip_credential
+-- Release Version 	: 1.2
+-- Purpose    		: Database Alter scripts for the release for credential DB.       
+-- Create By   		: Sadanandegowda DM
+-- Created Date		: Sep-2020
+-- 
+-- Modified Date        Modified By         Comments / Remarks
+-- -------------------------------------------------------------------------------------------------
+-- Apr-2021 		Ram Bhatt 	    status_comment column added.
+----------------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (ase-101) [reply db_release_scripts/mosip_credential/sql/1.2.1_release.sql:10]: Removed.

**Pull request merged**